src/grav/client.py

```diff
@@ -17,14 +17,44 @@
 
 
 class FriendlyHTTPResponse(HTTPResponse):
-    # I really like requests .json() method on response objects.
-    # this class is solely to enable that -- wdella 2019-10
+    """
+    A couple QoL improvements over http.client.HTTPResponse
+
+    I really like requests .json() method on response objects.
+
+    Also, in vanilla HTTPResponse objects, once read() there is no sanctioned
+    API to get the body content a second time.  Hope you remembed it! This
+    class allows accessing body content via .body, which lazy loads/caches the
+    body. Doing this is important to enable failUnlessStatus displaying body
+    info without having to worry about whether the test already read the body of
+    the response.
+    """
     def __init__(self, *args, **kwargs):
         super(FriendlyHTTPResponse, self).__init__(*args, **kwargs)
+        self._body = None
 
     def json(self):
         """Reads the response body, parses it as json, and returns the result."""
-        return json.loads(self.read())
+        # could be cached
+        return json.loads(self.body)
+
+    def read(self):
+        # clobber's the parent's read, which would be a problem if consumers
+        # expected regular HTTPResponse semantics.  However, I know that all
+        # the code in this repo accessing calls read without the amt arg.
+        #
+        # To do this the right way, I'd use an adapter pattern instead of
+        # inheritance. -- wdella 2019-10
+        data = super(FriendlyHTTPResponse, self).read()
+        if self._body is None:
+            self._body = data
```

Comment:
How would you do it if the response body was large? Not necessary for this project, I'm just curious.

Reply:
The right perf answer to this question largely depends on what we expect the size of the data to to be and what the expected read pattern. In this case the expected size is < 1Mb and the read pattern is once, maybe up to 3 times in quick succession, after which the response object will fall out of scope and get GC'd.  I think the current implementation is fine for that.

To entertain your question a bit further, imagine that we had response size `O(1/2 * available system memory)` that needs to be read many times over a long period.  I'll say a response less than 1/2 the system memory can be streamed out of the socket and into a io buffer a la https://github.com/javins/gravitational-audition/pull/5/commits/07d71b367bcdc606cba30858dc7b741c52a24adf. Greater than that and we'll persist it to disk (/tmp probably) and open the file back when needed to be read again.  This could also play nicely with logging, as we'd already have response data on persistent storage.

I recognize with swap you could probably get 'free' disk persistence by jamming the whole thing into memory, but that feels ugly to my gut.

Reply:
Let me know if you have other questions.  Until then, marking resolved.